### ISSUE TITLE
OpenIdConnectHandler: doesn't use updated user data from userInformationReceivedContext

### ISSUE THREAD

**Issue by almogohayon1** [dotnet/aspnetcore#18729]
### Describe the bug
Using OnUserInformationReceived event to override user information claims, after updating the user JsonElement provided in the context it's expected to be reflected in the issued identity.
After digging into the source code (https://github.com/dotnet/aspnetcore/blob/master/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs) i could see that the updated data element is ignored therefor not updated.

### To Reproduce
Use OnUserInformationReceived events to update the context.User data element with new/updated claim values.
Example: Add another claim type "sub" to context.User and make sure it's included in 
Options.ClaimsAction as DeleteClaimAction

- Expected: New user data to be used for running the ClaimActions on.
- Actual: New user data is ignored and "sub" claim is included.

### Source code to be updated

from
```
 foreach (var action in Options.ClaimActions)
  {
        action.Run(user.RootElement, identity, ClaimsIssuer);
  }
```
to
```
 foreach (var action in Options.ClaimActions)
  {
        action.Run(updatedUser.RootElement, identity, ClaimsIssuer);
  }
```

### Further technical details
- ASP.NET Core version
3.1.1
- Include the output of `dotnet --info`
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

**Comment by Tratcher** [comment]
This could be changed, but I've got to ask why you are modifying the JSON rather than directly modifying the ClaimsIdentity? I'd expect the ClaimsIdentity would be easier to work with.

**Comment by almogohayon1** [comment]
One example is to control duplicate claims for example "name" I would like to ensure only single claim from that type is processed. Obviously this claim is not yet populated to the identity at this phase, and if I'll add it eventually the identity will have more than one of those because of the claims actions running afterwards.

**Comment by anurse** [comment]
Duplicate claims should still end up in the ClaimsIdentity though, and you can filter them out from there rather than in the JSON.

**Comment by almogohayon1** [comment]
Duplicates should end up in the identity or not is up to the client to decide. OnUserInformationReceived is the place to control that behavior,  so I need control over response recieved from identity server, where in some cases where the response contains multiple claims from same time would end up showing as formatted array string mapped into the claimidentity claim name as follow "[name1, name2]" how do you suggest me to control that behavior?

**Comment by almogohayon1** [comment]
Any comments updates on my last response? is It going to be fixed?

**Comment by anurse** [comment]
> where in some cases where the response contains multiple claims from same time would end up showing as formatted array string mapped into the claimidentity claim name as follow "[name1, name2]" how do you suggest me to control that behavior?

So you're saying you're getting one claim with a single value that contains multiple items? Can you post a sample token and the way the claims are being loaded so we can get on the same page here?

**Comment by Tratcher** [comment]
> OnUserInformationReceived is the place to control that behavior

Another option is to set up some custom ClaimsActions that filter for duplicates as they add items.

**Comment by almogohayon1** [comment]
> > where in some cases where the response contains multiple claims from same time would end up showing as formatted array string mapped into the claimidentity claim name as follow "[name1, name2]" how do you suggest me to control that behavior?
> 
> So you're saying you're getting one claim with a single value that contains multiple items? Can you post a sample token and the way the claims are being loaded so we can get on the same page here?

Reproduction of the issue is very simple, to simulate it just call Run method in "UniqueJsonKeyClaimAction" and pass along JsonElement with array of claims from type "name" for example.
Because of the usage of JsonElement.GetString(..) you would get formatted string array.

(https://github.com/dotnet/aspnetcore/blob/master/src/Security/Authentication/OpenIdConnect/src/UniqueJsonKeyClaimAction.cs)

**Comment by Tratcher** [comment]
> and pass along JsonElement with array of claims from type "name" for example.

Like this?
```
{
   "name": ["value1", "value2"]
}
```

And you're rewriting it to this?
```
{
   "name": "value1"
}
```

**Comment by almogohayon1** [comment]
Yes

**Comment by Tratcher** [comment]
Here's an example that uses custom ClaimActions to map specific fields.
```
o.ClaimActions.DeleteClaim("name");
o.ClaimActions.MapCustomJson("name", "string", user =>
{
    var nameProperty = user.GetProperty("name");
    if (nameProperty.ValueKind == JsonValueKind.Array)
    {
        return nameProperty[0].GetString();
    }
    return nameProperty.GetString();
});
```

**Comment by almogohayon1** [comment]
Thank you,  I'm aware of that option already as a work around.. What I asked is to fix the bug..

**Comment by Tratcher** [comment]
A) Yes, this is an easy change, but since it's a minor issue and has a workaround, that fix won't come until the next release (5.0).
B) The ClaimAction seems preferable to rewriting the json, no? The Json APIs are designed for parsing, not modifying.
